### PR TITLE
[Draft] Dev Stats Log Writer

### PR DESCRIPTION
**Pull request by arjunsk**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [X] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

## What this PR does / why we need it:

### CODE REVIEW

File: pkg/fileservice/counter.go [hunk 1]
```diff
@@ -57,3 +65,23 @@ func WithCounter(ctx context.Context, counter *Counter) context.Context {
 	newCounters = append(newCounters, counter)
 	return context.WithValue(ctx, CtxKeyCounters, newCounters)
 }
+
+func (c *Counter) Collect() metric.Stats {
+	stats := make(map[string]int64)
+
+	c.Lock()
```
Comment (fengttt) [inline pkg/fileservice/counter.go:72]: why a lock is useful when writer does not lock?

Comment (arjunsk) [reply pkg/fileservice/counter.go:72]: I am assuming, we need to write a snapshot of all the Counter values.

File: pkg/fileservice/counter.go [hunk 2]
```diff
@@ -45,9 +50,12 @@ func updateCounters(ctx context.Context, fn func(*Counter)) {
 	}
 }
 
-func WithCounter(ctx context.Context, counter *Counter) context.Context {
+func WithCounter(ctx context.Context, statsFName string, counter *Counter) context.Context {
 	// check existed
 	v := ctx.Value(CtxKeyCounters)
+
+	metric.DefaultStatsRegistry.RegisterStats(statsFName, counter) //TODO: Need suggestions here.
```
Comment (fengttt) [inline pkg/fileservice/counter.go:57]: really?   you register with every call?

**Review by fengttt — CHANGES_REQUESTED** [review]
For each and every of the counter, you may need two entries,    total from system starts and counter within the collect period.      Total from system start will be used by moctl.

Lock is probably wrong, an atomic is actually easier.

I still feel putting all counters together is easier to use/understand.   But upto u.

**Comment by fengttt** [comment]
> For each and every of the counter, you may need two entries, total from system starts and counter within the collect period. Total from system start will be used by moctl.
> 
> Lock is probably wrong, an atomic is actually easier.
> 
> I still feel putting all counters together is easier to use/understand. But upto u.

Also in commit message, add how counters are added/used, and where to look for/search in system tables.   A few example would be helpful.   thank u.

File: pkg/fileservice/counter.go (hunk 2 above)
Comment (reusee) [inline pkg/fileservice/counter.go:57]: I think the `metric` package should implement its own WithCounter and call this WithCounter to set things up.

**Review by reusee — CHANGES_REQUESTED** [review]

File: pkg/fileservice/counter.go (hunk 1 above)
Comment (arjunsk) [reply pkg/fileservice/counter.go:72]: While Collecting, the initial state is 
```
NumRead = 5
NumHit = 5
```
We collected `NumRead=5`

Before collecting `NumHit=5`, `NumHit` is changed to `NumHit=7`

Now we collect `NumHit = 7`.

Should the output in the log be 
```
NumRead = 5
NumHit = 5
```
Or
```
NumRead = 5
NumHit = 7
```

File: pkg/fileservice/counter.go (hunk 2 above)
Comment (reusee) [reply pkg/fileservice/counter.go:57]: WithCounter may be called in every SQL statement or even every CN-DN request, the `metric` package should not register all of them.

Comment (arjunsk) [reply pkg/fileservice/counter.go:57]: Got it.

File: pkg/util/metric/mometric/stats_log_exporter.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	"context"
+	"github.com/matrixorigin/matrixone/pkg/logutil"
+	"github.com/matrixorigin/matrixone/pkg/util/metric"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type StatsLogExporter struct {
```
Comment (arjunsk) [inline pkg/util/metric/mometric/stats_log_exporter.go:12]: Global stats collector.

File: pkg/fileservice/counter.go (hunk 2 above)
Comment (arjunsk) [reply pkg/fileservice/counter.go:57]: Sorry. I will fix it.

**Comment by arjunsk** [comment]
Closing this Draft PR. The new PR https://github.com/matrixorigin/matrixone/pull/8431

**Pull request closed without merging**